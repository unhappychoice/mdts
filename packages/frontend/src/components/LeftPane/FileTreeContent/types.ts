export interface FileTreeItem {
  path: string;
  status: string;
}

export type FileTree = (FileTreeItem | { [key: string]: (FileTreeItem | object)[] })[];
