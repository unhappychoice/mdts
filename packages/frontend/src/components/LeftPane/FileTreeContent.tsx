import { ArticleOutlined, FolderOutlined } from '@mui/icons-material';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import { Box, CircularProgress, Typography, useTheme } from '@mui/material';
import { TreeItem } from '@mui/x-tree-view';
import { SimpleTreeView } from '@mui/x-tree-view/SimpleTreeView';
import React from 'react';
import { AppDispatch } from '../../store/store';

interface FileTreeItem {
  path: string;
  status: string;
}

interface FileTreeContentProps {
  filteredFileTree: (FileTreeItem | { [key: string]: (FileTreeItem | object)[] })[] | null;
  loading: boolean;
  error: string | null;
  expandedNodes: string[];
  onFileSelect: (path: string) => void;
  onExpandedItemsChange: (event: React.SyntheticEvent, itemIds: string[]) => void;
}

const FileTreeContent: React.FC<FileTreeContentProps> = ({
  filteredFileTree,
  loading,
  error,
  expandedNodes,
  onFileSelect,
  onExpandedItemsChange,
}) => {
  const theme = useTheme();

  const getStatusColor = (status: string) => {
    switch (status) {
      case 'M':
        return theme.palette.info.main;
      case '?':
      case 'A':
        return theme.palette.success.main;
      case 'D':
        return theme.palette.error.main;
      case 'R':
        return theme.palette.warning.main;
      case 'C':
        return theme.palette.secondary.main;
      default:
        return theme.palette.text.primary;
    }
  };

  const renderTreeItems = (
    tree: (FileTreeItem | { [key: string]: (FileTreeItem | object)[] })[] | null,
    onFileSelect: (path: string) => void,
    parentPath: string = ''
  ) => {
    if (!tree) {
      return null;
    }
    return tree.map((item) => {
      if ('path' in item) {
        const fileItem = item as FileTreeItem;
        const fileName = fileItem.path.split('/').pop();
        return (
          <TreeItem
            key={fileItem.path}
            itemId={fileItem.path}
            label={
              <Box sx={{ display: 'flex', alignItems: 'center' }}>
                <ArticleOutlined sx={{ mr: 1, fontSize: 'small' }} />
                <Typography variant="body2" sx={{ fontSize: '0.875rem', color: getStatusColor(fileItem.status) }}>
                  {fileName}
                </Typography>
                {fileItem.status && fileItem.status !== ' ' && (
                  <Typography variant="body2" sx={{ fontSize: '0.75rem', color: getStatusColor(fileItem.status), ml: 1, }}>
                    {fileItem.status}
                  </Typography>
                )}
              </Box>
            }
            onClick={() => onFileSelect(fileItem.path)}
          />
        );
      } else {
        const key = Object.keys(item)[0];
        const value = (item as any)[key];
        const currentPath = parentPath ? `${parentPath}/${key}` : key;
        return (
          <TreeItem
            key={currentPath}
            itemId={currentPath}
            label={
              <Box sx={{ display: 'flex', alignItems: 'center' }}>
                <FolderOutlined sx={{ mr: 1, fontSize: 'small' }} color="primary" />
                <Typography variant="body2" sx={{ fontSize: '0.875rem' }}>
                  {key}
                </Typography>
              </Box>
            }
          >
            {Array.isArray(value) && value.length > 0 && renderTreeItems(value, onFileSelect, currentPath)}
          </TreeItem>
        );
      }
    });
  };

  return (
    <>
      {loading ? (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: 'calc(100vh - 180px)' }}>
          <CircularProgress />
        </Box>
      ) : error ? (
        <Typography color="error">Error: {error}</Typography>
      ) : (
        <SimpleTreeView
          defaultCollapseIcon={<ExpandMoreIcon />}
          defaultExpandIcon={<ChevronRightIcon />}
          expandedItems={expandedNodes}
          onExpandedItemsChange={onExpandedItemsChange}
          sx={{ flexGrow: 1, maxWidth: 400, overflowY: 'auto', height: 'calc(100vh - 180px)' }}
        >
          {renderTreeItems(filteredFileTree, onFileSelect)}
        </SimpleTreeView>
      )}
    </>
  );
};

export default FileTreeContent;
