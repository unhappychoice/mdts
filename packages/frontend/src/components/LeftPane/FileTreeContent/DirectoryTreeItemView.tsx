import { FolderOutlined } from '@mui/icons-material';
import { Box, Typography } from '@mui/material';
import { TreeItem } from '@mui/x-tree-view';
import React from 'react';

interface DirectoryTreeItemViewProps {
  directoryName: string;
  children: React.ReactNode;
  currentPath: string;
}

export const DirectoryTreeItemView: React.FC<DirectoryTreeItemViewProps> = ({
  directoryName,
  children,
  currentPath
}) => (
  <TreeItem
    key={currentPath}
    itemId={currentPath}
    label={
      <Box sx={{ display: 'flex', alignItems: 'center' }}>
        <FolderOutlined sx={{ mr: 1, fontSize: 'small' }} color="primary" />
        <Typography variant="body2" sx={{ fontSize: '0.875rem' }}>
          {directoryName}
        </Typography>
      </Box>
    }
  >
    {children}
  </TreeItem>
);
