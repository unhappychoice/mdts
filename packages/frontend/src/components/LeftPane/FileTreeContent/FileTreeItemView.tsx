import { ArticleOutlined } from '@mui/icons-material';
import { Box, Typography } from '@mui/material';
import { TreeItem } from '@mui/x-tree-view';
import React, { useCallback } from 'react';
import { FileTreeItem } from './types';

interface FileTreeItemViewProps {
  fileItem: FileTreeItem;
  onFileSelect: (path: string) => void;
  getStatusColor: (status: string) => string;
}

export const FileTreeItemView: React.FC<FileTreeItemViewProps> = ({ fileItem, onFileSelect, getStatusColor }) => {
  const fileName = fileItem.path.split('/').pop();
  const handleClick = useCallback(() => onFileSelect(fileItem.path), [fileItem, onFileSelect]);

  return (
    <TreeItem
      key={fileItem.path}
      itemId={fileItem.path}
      label={
        <Box sx={{ display: 'flex', alignItems: 'center' }}>
          <ArticleOutlined sx={{ mr: 1, fontSize: 'small' }} />
          <Typography variant="body2" sx={{ fontSize: '0.875rem', color: getStatusColor(fileItem.status) }}>
            {fileName}
          </Typography>
          {fileItem.status && fileItem.status !== ' ' && (
            <Typography variant="body2" sx={{ fontSize: '0.75rem', color: getStatusColor(fileItem.status), ml: 1 }}>
              {fileItem.status}
            </Typography>
          )}
        </Box>
      }
      onClick={handleClick}
    />
  );
};
