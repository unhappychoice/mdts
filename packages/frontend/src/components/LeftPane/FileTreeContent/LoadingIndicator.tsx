import { Box, CircularProgress } from '@mui/material';
import React from 'react';

export const LoadingIndicator: React.FC = () => (
  <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: 'calc(100vh - 180px)' }}>
    <CircularProgress />
  </Box>
);
