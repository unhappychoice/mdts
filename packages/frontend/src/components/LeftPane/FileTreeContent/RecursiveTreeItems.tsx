import React from 'react';
import { DirectoryTreeItemView } from './DirectoryTreeItemView';
import { FileTreeItemView } from './FileTreeItemView';
import { FileTreeItem, FileTree } from './types';

interface RecursiveTreeItemsProps {
  tree: FileTree | null;
  onFileSelect: (path: string) => void;
  parentPath?: string;
  getStatusColor: (status: string) => string;
}

export const RecursiveTreeItems: React.FC<RecursiveTreeItemsProps> = ({ tree, onFileSelect, parentPath = '', getStatusColor }) => {
  if (!tree) return null;

  return tree.map((item) => {
    if ('path' in item) {
      return <FileTreeItemView
        key={item.path}
        fileItem={item as FileTreeItem}
        getStatusColor={getStatusColor}
        onFileSelect={onFileSelect} />;
    } else {
      const key = Object.keys(item)[0];
      const value = item[key];
      const currentPath = parentPath ? `${parentPath}/${key}` : key;
      return (
        <DirectoryTreeItemView key={key} directoryName={key} currentPath={currentPath}>
          {Array.isArray(value) && value.length > 0 && (
            <RecursiveTreeItems
              tree={value}
              parentPath={currentPath}
              getStatusColor={getStatusColor}
              onFileSelect={onFileSelect} />
          )}
        </DirectoryTreeItemView>
      );
    }
  });
};
