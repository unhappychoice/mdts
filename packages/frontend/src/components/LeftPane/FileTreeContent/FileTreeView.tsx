import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import { SimpleTreeView } from '@mui/x-tree-view/SimpleTreeView';
import React from 'react';
import { RecursiveTreeItems } from './RecursiveTreeItems';
import { FileTree } from './types';

interface FileTreeViewProps {
  filteredFileTree: FileTree | null;
  expandedNodes: string[];
  onFileSelect: (path: string) => void;
  onExpandedItemsChange: (event: React.SyntheticEvent, itemIds: string[]) => void;
  getStatusColor: (status: string) => string;
}

export const FileTreeView: React.FC<FileTreeViewProps> = ({
  filteredFileTree,
  expandedNodes,
  onFileSelect,
  onExpandedItemsChange,
  getStatusColor
}) => (
  <SimpleTreeView
    className="custom-scrollbar"
    defaultCollapseIcon={<ExpandMoreIcon />}
    defaultExpandIcon={<ChevronRightIcon />}
    expandedItems={expandedNodes}
    onExpandedItemsChange={onExpandedItemsChange}
    sx={{ flexGrow: 1, maxWidth: 400, overflowY: 'auto', height: 'calc(100vh - 180px)', px: 2, pb: 2 }}
  >
    <RecursiveTreeItems tree={filteredFileTree} onFileSelect={onFileSelect} getStatusColor={getStatusColor} />
  </SimpleTreeView>
);
