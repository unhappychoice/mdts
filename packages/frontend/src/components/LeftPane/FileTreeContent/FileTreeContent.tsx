import { useTheme } from '@mui/material';
import React, { useCallback } from 'react';
import ErrorView from '../../ErrorView';
import { FileTreeView } from './FileTreeView';
import { LoadingIndicator } from './LoadingIndicator';
import { FileTree } from './types';

interface FileTreeContentProps {
  filteredFileTree: FileTree | null;
  loading: boolean;
  error: string | null;
  expandedNodes: string[];
  onFileSelect: (path: string) => void;
  onExpandedItemsChange: (event: React.SyntheticEvent, itemIds: string[]) => void;
}

const FileTreeContent: React.FC<FileTreeContentProps> = ({
  filteredFileTree,
  loading,
  error,
  expandedNodes,
  onFileSelect,
  onExpandedItemsChange,
}) => {
  const theme = useTheme();

  const getStatusColor = useCallback((status: string) => {
    switch (status) {
      case 'M': return theme.palette.info.main;
      case '?':
      case 'A': return theme.palette.success.main;
      case 'D': return theme.palette.error.main;
      case 'R': return theme.palette.warning.main;
      case 'C': return theme.palette.secondary.main;
      default: return theme.palette.text.primary;
    }
  }, [theme]);

  if (loading && !(filteredFileTree && filteredFileTree.length > 0)) return <LoadingIndicator />;
  if (error) return <ErrorView error={error} />;

  return (
    <FileTreeView
      filteredFileTree={filteredFileTree}
      expandedNodes={expandedNodes}
      onFileSelect={onFileSelect}
      onExpandedItemsChange={onExpandedItemsChange}
      getStatusColor={getStatusColor}
    />
  );
};

export default FileTreeContent;
