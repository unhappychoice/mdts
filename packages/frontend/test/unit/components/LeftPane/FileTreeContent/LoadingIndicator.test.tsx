import { render } from '@testing-library/react';
import React from 'react';
import { LoadingIndicator } from '../../../../../src/components/LeftPane/FileTreeContent/LoadingIndicator';

describe('LoadingIndicator', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<LoadingIndicator />);
    expect(asFragment()).toMatchSnapshot();
  });
});
