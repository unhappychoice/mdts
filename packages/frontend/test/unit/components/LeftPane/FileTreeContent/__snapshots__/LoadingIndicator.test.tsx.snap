// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LoadingIndicator should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-hehwu5"
  >
    <span
      class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;
